openapi_python_client/templates/property_templates/union_property.pyi

```diff
@@ -9,6 +9,17 @@ def _parse_{{ property.python_name }}(data: {{ property.get_type_string(json=Tru
     if isinstance(data, Unset):
         return data
     {% endif %}
+    {% if property.discriminator_property != None %}
+    discriminator_value: str = data.get("{{property.discriminator_property}}")
+    if discriminator_value is not None:
+        {% for discriminated_by, inner_property in property.discriminator_mappings.items() %}
+        if discriminator_value == "{{discriminated_by}}":
+            {% from "property_templates/" + inner_property.template import construct %}
+            {{ construct(inner_property, "data", initial_value="UNSET") | indent(8) }}
+            return {{ property.python_name }}
+        {% endfor %}
+
+    {% endif %}
```

Comment:
I originally wrote this to fall back nicely to the brute force approach if a discriminator is accidentally missing, but I think it's probably more correct to have that be enforced by the openapi linter and raise an error at the end of this if chain here if we haven't returned.

Reply:
Likewise, William had pointed out that we might want to have a default value for types for forward compatibility, but the existing mechanism doesn't do that, so we have to decide whether to add it for non-discriminated cases as well.  I'm still thinking this through a little bit, but given that the UnionProperty type can also be used for "simple" types like `str` vs `int` (which can never be discriminated, if I'm reading the OpenAPI docs correctly), best case is the logic gets strange based on the types that can be unioned here.

I'm tempted to take a stand and say "if you want to support forward compatibility, use discriminators" and implement the "Unknown" case for that only.

Reply:
I'll need to return to this tomorrow after I've thought about it some, too.

But I think forward compatibility is the key consideration here. And the problems get only more acute the deeper a prop is embedded in the model hierarchy.

We could enforce your rule on discriminators for forward compatibility internally, but that might be a hard sell if we're trying to upstream this.

Reply:
> best case is the logic gets strange based on the types that can be unioned here.

Can't we detect this from the value we get (as opposed to the spec)? Like if it's a primitive (i.e. just not a `dict` or `list` when serialized), then we don't do any deserialization - the raw value is the deserialzed value. If it is a dict, then we do the existing brute force approach (since in this case it'd be models union'ed with primitives for some reason and we can discriminate)